servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -154,57 +166,118 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
-                            return newHosts;
+                        } else if (event.isAvailable()) {
+                            return addHostToList(oldHostsTyped, addr, true);
                         } else if (oldHostsTyped.isEmpty()) {
                             return emptyList();
                         } else {
-                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() - 1);
-                            for (int i = 0; i < oldHostsTyped.size(); ++i) {
-                                final Host<ResolvedAddress, C> host = oldHostsTyped.get(i);
-                                if (host.address.equals(addr)) {
-                                    host.markInactive();
-                                    for (int x = i + 1; x < oldHostsTyped.size(); ++x) {
-                                        newHosts.add(oldHostsTyped.get(x));
-                                    }
-                                    return newHosts.isEmpty() ? emptyList() : newHosts;
-                                } else {
-                                    newHosts.add(host);
-                                }
-                            }
-                            return newHosts;
+                            return markHostAsExpired(oldHostsTyped, addr);
                         }
                     });
 
                 LOGGER.debug("Load balancer {} now using {} addresses: {}", RoundRobinLoadBalancer.this,
-                        activeAddresses.size(), activeAddresses);
+                        usedAddresses.size(), usedAddresses);
 
                 if (event.isAvailable()) {
-                    if (activeAddresses.size() == 1) {
+                    if (usedAddresses.size() == 1) {
                         eventStreamProcessor.onNext(LOAD_BALANCER_READY_EVENT);
                     }
-                } else if (activeAddresses.isEmpty()) {
+                } else if (usedAddresses.isEmpty()) {
                     eventStreamProcessor.onNext(LOAD_BALANCER_NOT_READY_EVENT);
                 }
             }
 
+            private List<Host<ResolvedAddress, C>> markHostAsExpired(
+                    final List<Host<ResolvedAddress, C>> oldHostsTyped, final ResolvedAddress addr) {
+                for (Host<ResolvedAddress, C> host : oldHostsTyped) {
+                    if (host.address.equals(addr)) {
+                        // Host removal will be handled by the Host's onClose::afterFinally callback
+                        host.markExpired();
+                    }
+                }
+                return oldHostsTyped;
+            }
+
+            @SuppressWarnings("unchecked")
+            private Host<ResolvedAddress, C> createHost(ResolvedAddress addr) {
+                Host<ResolvedAddress, C> host = new Host<>(addr);
+                if (!eagerConnectionShutdown) {
+                    host.onClose().afterFinally(() -> {
```

Comment:
replace lambda with expression.

also consider reducing scope of `@SuppressWarnings("unchecked")` to site of warning. this can help if the method evolves to avoid unintended warnings from being suppressed.

```java
previousHosts -> {
  @SuppressWarnings("unchecked")
  List<Host<ResolvedAddress, C>> cast = (List<Host<ResolvedAddress, C>>) previousHosts;
  return listWithHostRemoved(cast, current -> current == host);
})
```